### ISSUE TITLE
[Discuss]should onHasTask being notified by `periodMacroTask`

### ISSUE THREAD

**Issue by JiaLiPassion** [angular/zone.js#1017]
based on #999, current `setInterval` 's behavior is 
- when call `setInterval` will schedule a macroTask with `isPeriodic=true`, and trigger `onHasTask` with `macroTask count +1`.
- and when `interval callback` is invoked, onHasTask will not be invoked because it is `periodic`.
- when `cancelInterval` is called, will trigger `onHasTask` with `macroTask count -1`. 

It is the correct behavior.

But in `onHasTask` callback, we can only get 

```javascript
type HasTaskState = {
  microTask: boolean; macroTask: boolean; eventTask: boolean; change: TaskType;
};
```

`macroTask` is true or false, but we can not get whether the `macroTask` isPeriodic or not.
In some case, we may want to consider 
- all microTasks
- all non-periodic macroTasks

are stable.

For example, in some application, we start a `interval` to polling from server at background, and basically we should ignore this `interval` task when we want to check the application is `stable` or not. otherwise the application will not stable at all.

@mhevery , do you think we need to provide something like 

```javascript
type HasTaskState = {
  microTask: boolean; 
  macroTask: boolean; 
  eventTask: boolean;
  change: TaskType;
  isAllMacroTasksPeriodic: boolean;
};
```
to provide `isPeriodic` information and not making breaking changes?

Please review, thank you!

**Comment by mhevery** [comment]
What is the need to know if there are periodic tasks? I think the answers should be just to run the periodicTask outside the current zone so that it does not show up here.

**Comment by JiaLiPassion** [comment]
@mhevery , the reason about this question is https://github.com/angular/angular/issues/20970, 
`service worker` will not register because `ngZone` will never stable when `setInetrval` is called.

run `setInterval` outsideAngular will cause the `callback` outside of `ngZone`, so the walk around is 

```javascript
this.ngZone.runOutsideAngular(() => {
  setInterval(() => {
    this.ngZone.run(() => {
        // do something
    });
  }, 1000);
});
```

to make the `setInterval` not trigger `onHasTask` of `ngZone` but can also run `callback` inside `ngZone`.

So I think maybe a `isPeriod` option may help? such as 

```javascript
this.ngZone.isStable.subscrible((taskStates) => {
  if (taskStates.isAllPeriodTasks || !(taskStates.hasPendingMicroTasks || taskState.hasPendingMacroTasks) {
    // consider ngZone is stable
  }
});
```

or we can provide some helper method to make this simple wrap version(no task) of `setInterval` more easier

```javascript
this.ngZone.runOutsideAngular(() => {
  setInterval(() => {
    this.ngZone.run(() => {
        // do something
    });
  }, 1000);
});
```

something like `setIntervalWrap()`...?

**Comment by mhevery** [comment]
I feel that your proposal is a fix to a specific issue of angular/angular#20970 rather than a general need. `setInterval` is no different from `work = () => setInterval(work)` and in that case the issue would persist. If there is outstanding work than zone will point this out. To special case long lived work just feels weird since it may be that that work is needed. For example animation may temporarily create `setInterval` only to cancel it later.

**Comment by JiaLiPassion** [comment]
@mhevery , got it, thank you. I agree this is not a `general need`, and should not be handled in `zone.js`.  I will try to add it to documentation.
